### DISCUSSION TITLE
extract response type from the handler api function in app router

### DISCUSSION THREAD

**Discussion by Xenos112** [vercel/next.js#68796] [App Router] [answered]
i have noticed that there is a type for the api handler, the type contains the response schema, is there a way to extract it
![Code_W5Ch9lUq5C](https://github.com/user-attachments/assets/9943dced-497b-4456-8a53-87464828bd8a)

**Comment by icyJoseph** [comment]
Yeah, like:


```ts
type Foo = Awaited<Return type<type of GET>>
```

That'll give you the return type, wrapped by NextResponse, and from there you could use a helper to infer the result :) want that code too?

**Reply by Xenos112** [reply]
This does not work, it returns what the `NextResponse` have like json and url

**Reply by icyJoseph** [reply, marked as answer]
I tried to, not take away, an opportunity to work it out yourself, but here it goes:

```ts
type ExtractFromNextResponse<Result> = Result extends NextResponse<
  infer Payload
>
  ? Payload
  : never;

type GETTypeReturn = ExtractFromNextResponse<Awaited<ReturnType<typeof GET>>>;

```
